### PR TITLE
Introduce new AdapterIdFragment to replace CustomData usage

### LINKED ISSUE #255: Adapter id stored in Fragments
#### Description:

As discussed in various occasions, including https://github.com/BHoM/BHoM_Adapter/issues/167

we are going to introduce a new fragment interface
`IAdapterId`
to be placed in the Base BHoM. This interface will be for a class that has to host an id object.
We can have another look about generics here, but as we saw in some tests in the past it might be better to just refer to `object` or `string` to simplify the framework and allow the parsing to happen in the specific toolkit.

### PR DESCRIPTION
**Pull request by alelom**
### NOTE: Depends on 

https://github.com/BHoM/BHoM/pull/1033
   
### Issues addressed by this PR

Closes #255 

### Test files

See individual adapters.

### CODE REVIEW

**Comment by peterjamesnugent** [comment]
@alelom question on this.

Should all the methods be `AdapterIdName` instead of `AdapterId`, we did make a change a while back to use `AdapterNameId` in place of `AdapterId`.

Or are we now reverting that change? May be worth clarifying before `CustomData` and hence the use of `AdapterIdName` is removed from the various adapters.

**Comment by alelom** [comment]
> @alelom question on this.
> 
> Should all the methods be `AdapterIdName` instead of `AdapterId`, we did make a change a while back to use `AdapterNameId` in place of `AdapterId`.
> 
> Or are we now reverting that change? May be worth clarifying before `CustomData` and hence the use of `AdapterIdName` is removed from the various adapters.

Sorry, I should've reverted to Draft for this PR - still revising some details. I will reach you guys out for explanation today in the Structures call for more details.

File: Adapter_Engine/Modify/SetAdapterId.cs [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    public static partial class Modify
+    {
+        /***************************************************/
+        /**** Public Methods                            ****/
+        /***************************************************/
+
+        public static void SetAdapterId<T>(this IBHoMObject bHoMObject, Type adapterIdFragmentType, T id)
+        {
+            // Check if the specified `adapterIdFragmentType` is effectively an `IAdapterId`.
+            if (typeof(IAdapterId).IsAssignableFrom(adapterIdFragmentType))
```
Comment (IsakNaslundBh) [inline Adapter_Engine/Modify/SetAdapterId.cs:44]:
```suggestion
            if (!typeof(IAdapterId).IsAssignableFrom(adapterIdFragmentType))
```

File: Adapter_Engine/Query/AdapterId.cs [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+{
+    public static partial class Query
+    {
+        /***************************************************/
+        /**** Public Methods                            ****/
+        /***************************************************/
+
+        public static object AdapterId(this IBHoMObject bHoMObject, Type adapterIdFragmentType)
+        {
+            if (typeof(IAdapterId).IsAssignableFrom(adapterIdFragmentType))
```
Comment (IsakNaslundBh) [inline Adapter_Engine/Query/AdapterId.cs:44]:
```suggestion
            if (!typeof(IAdapterId).IsAssignableFrom(adapterIdFragmentType))
```

**Review by IsakNaslundBh — CHANGES_REQUESTED** [review]
Some logic switches, see below:

File: Adapter_Engine/Query/AdapterId.cs [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+            {
+                BH.Engine.Reflection.Compute.RecordError($"The `{adapterIdFragmentType.Name}` is not a valid `{typeof(IAdapterId).Name}`.");
+                return null;
+            }
+
+            List<IAdapterId> fragmentList = bHoMObject.GetAllFragments(adapterIdFragmentType).OfType<IAdapterId>().ToList();
+
+            if (fragmentList.Count != 0)
+            {
+                IEnumerable<object> ids = fragmentList.Select(f => IAdapterId(f as IAdapterId));
```
Comment (IsakNaslundBh) [inline Adapter_Engine/Query/AdapterId.cs:54]:
```suggestion
                IEnumerable<object> ids = fragmentList.Select(f => f.Id);
```

File: Adapter_Engine/Query/AdapterId.cs [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+
+        private static object IAdapterId(IAdapterId adapterIdFragment)
+        {
+            return AdapterId(adapterIdFragment as dynamic);
+        }
+
+        private static object AdapterId(IAdapterId adapterId)
+        {
+            return Engine.Reflection.Compute.RunExtensionMethod(adapterId, "IAdapterId");
+        }
```
Comment (IsakNaslundBh) [inline Adapter_Engine/Query/AdapterId.cs:92]:
```suggestion

```

File: Adapter_Engine/Modify/SetAdapterId.cs [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+            // Check if the specified `adapterIdFragmentType` is effectively an `IAdapterId`.
+            if (!typeof(IAdapterId).IsAssignableFrom(adapterIdFragmentType))
+            {
+                BH.Engine.Reflection.Compute.RecordError($"The `{adapterIdFragmentType.Name}` is not a valid `{typeof(IAdapterId).Name}`.");
+                return;
+            }
+
+            // If the input `id` is already an instance of the specified `adapterIdFragmentType`, simply add it to the fragments.
+            if (typeof(T) == adapterIdFragmentType)
+                bHoMObject.Fragments.AddOrReplace(id as IFragment);
```
Comment (IsakNaslundBh) [inline Adapter_Engine/Modify/SetAdapterId.cs:52]: Cant trust the type of `T` here. For test case with etabs it was object, hence could not be relied upon.

Also need to exit the method if this is true
```suggestion
            if (id.GetType() == adapterIdFragmentType)
            {
                bHoMObject.Fragments.AddOrReplace(id as IFragment);
                return;
            }
```

**Review by IsakNaslundBh — CHANGES_REQUESTED** [review]
SetAdapterId issues, see below

File: Adapter_Engine/Modify/SetAdapterId.cs [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+
+            // Create an instance of the specified `adapterIdFragmentType`, set its `Id` property, then add it to the fragments.
+            IAdapterId newAdapterIdFragment = (IAdapterId)Activator.CreateInstance(adapterIdFragmentType);
+
+            var idProp = newAdapterIdFragment.GetType().GetProperty(nameof(BH.oM.Base.IAdapterId.Id));
+            if (idProp != null)
+                idProp.SetValue(newAdapterIdFragment, id, null);
```
Comment (IsakNaslundBh) [inline Adapter_Engine/Modify/SetAdapterId.cs:59]: Also, no real reason to be using reflection to set the property here. Could just change this to

```suggestion
            newAdapterIdFragment.Id = id;
```

File: Adapter_Engine/Modify/SetAdapterId.cs (hunk 5 above)
Comment (alelom) [reply Adapter_Engine/Modify/SetAdapterId.cs]: Good catch, thanks!

File: Adapter_Engine/Modify/SetAdapterId.cs (hunk 6 above)
Comment (alelom) [reply Adapter_Engine/Modify/SetAdapterId.cs]: Right, this is a leftover from when we were relying on Generics. Thanks.

**Review by IsakNaslundBh — DISMISSED** [review]
This all seem to be working as expected now!

File: Adapter_Engine/Modify/SetAdapterId.cs [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+        {
+            // Check if the specified `adapterIdFragmentType` is effectively an `IAdapterId`.
+            if (!typeof(IAdapterId).IsAssignableFrom(adapterIdFragmentType))
+            {
+                BH.Engine.Reflection.Compute.RecordError($"The `{adapterIdFragmentType.Name}` is not a valid `{typeof(IAdapterId).Name}`.");
+                return;
+            }
+
+            // If the input `id` is already an instance of the specified `adapterIdFragmentType`, simply add it to the fragments.
+            if (id.GetType() == adapterIdFragmentType)
```
Comment (IsakNaslundBh) [inline Adapter_Engine/Modify/SetAdapterId.cs:51]:
```suggestion
            if (id != null && id.GetType() == adapterIdFragmentType)
```

File: BHoM_Adapter/AdapterActions/_PushMethods/AdapterId/AssignId.cs [hunk 8]
```diff
@@ -39,9 +41,9 @@ public abstract partial class BHoMAdapter
             bool refresh = true;
             foreach (T item in objects)
             {
-                if (AdapterIdName != null)
+                if (AdapterIdFragmentType != null)
                 {
-                    item.CustomData[AdapterIdName] = NextFreeId(typeof(T), refresh);
+                    item.SetAdapterId(AdapterIdFragmentType, NextFreeId(typeof(T), refresh));
                     refresh = false;
```
Comment (IsakNaslundBh) [inline BHoM_Adapter/AdapterActions/_PushMethods/AdapterId/AssignId.cs:47]: Extra null change here:

```suggestion
                    object nextId = NextFreeId(typeof(T), refresh);
                    if (nextId != null)
                    {
                        item.SetAdapterId(AdapterIdFragmentType, nextId);
                        refresh = false;
                    }
```

**Review by IsakNaslundBh — COMMENTED** [review]
Some extra null checks for setting and assigning adapter Id

**Review by IsakNaslundBh — APPROVED** [review]
Approving again!

**Pull request merged**